src/stdlib_strings.f90

```diff
@@ -649,4 +660,87 @@ pure function replace_all_char_char_char(string, pattern, replacement) result(re
 
     end function replace_all_char_char_char
 
+    !> Returns the number of times substring 'pattern' has appeared in the
+    !> input string 'string'
+    !> Returns an integer
+    elemental function count_string_string(string, pattern, consider_overlapping) result(res)
+        type(string_type), intent(in) :: string
+        type(string_type), intent(in) :: pattern
+        logical, intent(in), optional :: consider_overlapping
+        integer :: res
+
+        res = count(char(string), char(pattern), consider_overlapping)
+
+    end function count_string_string
+
+    !> Returns the number of times substring 'pattern' has appeared in the
+    !> input string 'string'
+    !> Returns an integer
+    elemental function count_string_char(string, pattern, consider_overlapping) result(res)
+        type(string_type), intent(in) :: string
+        character(len=*), intent(in) :: pattern
+        logical, intent(in), optional :: consider_overlapping
+        integer :: res
+
+        res = count(char(string), pattern, consider_overlapping)
+
+    end function count_string_char
+
+    !> Returns the number of times substring 'pattern' has appeared in the
+    !> input string 'string'
+    !> Returns an integer
+    elemental function count_char_string(string, pattern, consider_overlapping) result(res)
+        character(len=*), intent(in) :: string
+        type(string_type), intent(in) :: pattern
+        logical, intent(in), optional :: consider_overlapping
+        integer :: res
+
+        res = count(string, char(pattern), consider_overlapping)
+
+    end function count_char_string
+
+    !> Returns the number of times substring 'pattern' has appeared in the
+    !> input string 'string'
+    !> Returns an integer
+    elemental function count_char_char(string, pattern, consider_overlapping) result(res)
+        character(len=*), intent(in) :: string
+        character(len=*), intent(in) :: pattern
+        logical, intent(in), optional :: consider_overlapping
+        integer :: lps_array(len(pattern))
+        integer :: res, s_i, p_i, length_string, length_pattern
+        logical :: consider_overlapping_
+
+        consider_overlapping_ = optval(consider_overlapping, .true.)
+        res = 0
+        length_string = len(string)
+        length_pattern = len(pattern)
+
+        if (length_pattern > 0 .and. length_pattern <= length_string) then
+            lps_array = compute_lps(pattern)
+
+            s_i = 1
+            p_i = 1
+            do while(s_i <= length_string)
+                if (string(s_i:s_i) == pattern(p_i:p_i)) then
+                    if (p_i == length_pattern) then
+                        res = res + 1
+                        if (consider_overlapping_) then
```

Comment:
`consider_overlapping_` is used only once. So we may as well take advantage of `optval()` to full effect and drop the temporary variable `consider_overlapping_`:

```suggestion

        res = 0
        length_string = len(string)
        length_pattern = len(pattern)

        if (length_pattern > 0 .and. length_pattern <= length_string) then
            lps_array = compute_lps(pattern)

            s_i = 1
            p_i = 1
            do while(s_i <= length_string)
                if (string(s_i:s_i) == pattern(p_i:p_i)) then
                    if (p_i == length_pattern) then
                        res = res + 1
                        if (optval(consider_overlapping, .true.)) then
```

Reply:
done!